Airpower rework

troyaf opened the pull request:
Closes #74 

- [x] Document exported command
- [x] Add markdown documentation of CLI with examples
- [x] Replace use of intermediate tar file with gzip custom seek
- [x] Extraction progress async
- [x] Add more tests
- [x] Add coverage to workflow
- [x] Rework db to use the filesystem
- [x] Enable `exec` command to use local packages for testing/shipyard
- [x] Perform an HTTP HEAD on `*/manifests/*` endpoints before pulling to better avoid hitting the docker api request limit.

tbateson commented on .github/workflows/test.yml:
```diff
@@ -9,5 +9,4 @@ jobs:
-      run: |
-        cd test; ./run.ps1
\ No newline at end of file
+      run: .\test.ps1
```
```suggestion
      run: .\test.ps1

```

tbateson commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+.modules
+build
+*.xml
```
```suggestion
*.xml

```

tbateson commented on src/progress.ps1:
```diff
@@ -0,0 +1,116 @@
+# U+2588	█	Full block
+# U+2589	▉	Left seven eighths block
+# U+258A	▊	Left three quarters block
+# U+258B	▋	Left five eighths block
+# U+258C	▌	Left half block
+# U+258D	▍	Left three eighths block
+# U+258E	▎	Left one quarter block
+# U+258F	▏	Left one eighth block
```
Some of these characters do not render properly. My recommendation is we only use the full block, half block, and empty block.

nicklauslittle-gov replied:
I haven't looked into this at all, but there is a PowerShell function available: [Write-Progress](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/write-progress). I'm guessing it is insufficient for whatever we are doing?

tbateson replied:
It's insufficient in the sense that the aesthetics of it looks bad.

tbateson replied:
@troyaf Should we remove the unicode values here that we aren't using?

troyaf replied:
Done

tbateson commented on src/config.ps1:
```diff
@@ -0,0 +1,95 @@
+function GetPwrDB {
+	$db = GetPwrDBPath
+	if (Test-Path $db -PathType Leaf) {
+		Get-Content $db -Raw | ConvertFrom-Json | ConvertTo-HashTable
```
This really shouldn't be human-readable or editable.

troyaf replied:
Resolved

tbateson commented on src/registry.ps1:
```diff
@@ -0,0 +1,119 @@
+		$task = $resp.Content.CopyToAsync($fs)
+		while (-not $task.IsCompleted) {
+			$sha256.Substring(0,12) + ': Downloading ' + (GetProgress -Current $fs.Length -Total $size) + '  ' | WriteConsole
+			Start-Sleep -Milliseconds 17
```
```suggestion
			Start-Sleep -Milliseconds 125
```

tbateson commented on src/tar.ps1:
```diff
@@ -0,0 +1,159 @@
+		$task = $gz.CopyToAsync($stream)
+		while (-not $task.IsCompleted) {
+			$layer.Substring(0,12) + ': Decompressing ' + (GetProgress -Current $fs.Position -Total $fs.Length) | WriteConsole
+			Start-Sleep -Milliseconds 17
```
```suggestion
			Start-Sleep -Milliseconds 125
```

tbateson commented on src/config.ps1:
```diff
@@ -0,0 +1,95 @@
+	} elseif ($env:PwrHome) {
+		$env:PwrHome
+	} else {
+		"$env:LocalAppData\pwr"
```
Since we're moving this location, should we take the opportunity to rename to `$env:LocalAppData\Airpower`?

tbateson commented on src/config.ps1:
```diff
@@ -0,0 +1,95 @@
+}
+
+function GetPwrDBPath {
+	"$(GetPwrHome)\pwrdb"
```
Same sorta question as above, why not just name this `$(GetPwrHome)\db`?

troyaf replied:
ok

troyaf replied:
Done

tbateson commented on src/config.ps1:
```diff
@@ -0,0 +1,95 @@
+}
+
+function GetPwrTempPath {
+	"$(GetPwrHome)\tmp"
```
I prefer `temp`, you even use it in the function name.

```suggestion
	"$(GetPwrHome)\temp"
```

tbateson commented on src/config.ps1:
```diff
@@ -0,0 +1,95 @@
+	if ($PwrHome) {
+		$PwrHome
+	} elseif ($env:PwrHome) {
+		$env:PwrHome
```
Same vein as below, do we want this to be `$env:AirpowerHome`?

tbateson replied:
What about `$env:AirpowerPath`, similar to `$env:GoPath`, since the command is not in this directory anymore, doesn't seem correct to call it the home.

troyaf replied:
I like `$env:AirpowerPath`

troyaf replied:
Done

tbateson commented on src/registry.ps1:
```diff
@@ -0,0 +1,119 @@
+		$size = $resp.Content.Headers.ContentLength + $fs.Length
+		$task = $resp.Content.CopyToAsync($fs)
+		while (-not $task.IsCompleted) {
+			$sha256.Substring(0,12) + ': Downloading ' + (GetProgress -Current $fs.Length -Total $size) + '  ' | WriteConsole
```
```suggestion
			$sha256.Substring(0, 12) + ': Downloading ' + (GetProgress -Current $fs.Length -Total $size) + '  ' | WriteConsole
```

tbateson commented on src/tar.ps1:
```diff
@@ -0,0 +1,172 @@
+			}
+			$leftover = $size % 512
+			if ($leftover -gt 0) {
+				$stream.Seek(512 - ($size % 512), [IO.SeekOrigin]::Current) | Out-Null
```
```suggestion
				$stream.Seek(512 - $leftover, [IO.SeekOrigin]::Current) | Out-Null
```

troyaf commented on src/progress.ps1:
```diff
@@ -0,0 +1,116 @@
+	if (($null -eq $lastwrite) -or (((get-date) - $lastwrite) -gt 125)) {
+		[Console]::Write("`r$Line")
+		$script:lastwrite = (get-date)
+	}
```
Done

nicklauslittle commented:
A couple things I've noticed using this version (not sure any of these are blockers for #74, but maybe issues after v0.6 is available:
- Install seems a lot slower than it used to be
- Noticed that after downloading, the tar.gz is decompressed into another temp file, which could result in excessive disk usage with larger packages (e.g., vs-buildtools would use ~4GB for tar.gz, ~15GB for .tar, ~15GB for content)

troyaf commented:
> A couple things I've noticed using this version (not sure any of these are blockers for #74, but maybe issues after v0.6 is available:
> 
> * Install seems a lot slower than it used to be
> * Noticed that after downloading, the tar.gz is decompressed into another temp file, which could result in excessive disk usage with larger packages (e.g., vs-buildtools would use ~4GB for tar.gz, ~15GB for .tar, ~15GB for content)

.NET added a TAR class in version 7, but since we're trying to be 5.1 compatible without calling system tar, we're left with implementing it directly in PowerShell. We can remove the intermediate tar file by using the gzip stream directly; however, I noticed that interrupting the extraction corrupts the gzip base stream, so if we use the tgz directly, it will result in a redownload. Also, an asynchronous untar will need to have random-ish access to the underlying tar bytes, otherwise we're left decompressing our way to the right position. Alternatively, we can use an internal memory buffer to avoid multiple passes and spawn multiple threads to extract files in small batches.

I'm weighing the options. It is slow, and I agree it needs to be faster.

nicklauslittle-gov commented:
Seeing some strange issues on AFNet. Install works fine, but when I try to pull I get strange results:

```pwsh
PS C:\Users\1379838568C> Airpower version

Major  Minor  Build  Revision
-----  -----  -----  --------
0      6      0      9


PS C:\Users\1379838568C> Airpower pull vs-buildtools:17.4.3
17.4.3: Pulling vs-buildtools
Digest: sha256:b8958be9f66a63e705501ceec1bb2fe7f6f267ec947ddc1016994df66e7841cf

HttpSend : Exception calling "GetResult" with "0" argument(s): "A task was canceled."
At C:\Users\1379838568C\Documents\WindowsPowerShell\Modules\Airpower\0.6.0.9\Airpower.psm1:490 char:31
+     return HttpRequest @params | HttpSend
+                                  ~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [HttpSend], MethodInvocationException
    + FullyQualifiedErrorId : TaskCanceledException,HttpSend

PS C:\Users\1379838568C> Airpower pull python
latest: Pulling python
Digest: sha256:2f59a0baa4215044bea41a569baba96ad575ff9af75409584233a784d520cbc8
095b3609a830: Downloading █████████████▌                   17.50 MB / 38.2 MB
removed C:\Users\1379838568C\AppData\Local\Airpower\temp\095b3609a830900f1853c74f05445f78a7deb66b2622814ecdccd785f9506f5e.tar.gz because it had
corrupted data
At C:\Users\1379838568C\Documents\WindowsPowerShell\Modules\Airpower\0.6.0.9\Airpower.psm1:372 char:3
+         throw "removed $Path because it had corrupted data"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (removed C:\User... corrupted data:String) [], RuntimeException
    + FullyQualifiedErrorId : removed C:\Users\1379838568C\AppData\Local\Airpower\temp\095b3609a830900f1853c74f05445f78a7deb66b2622814ecdccd785f9506
   f5e.tar.gz because it had corrupted data
```

When I try to pull `vs-buildtools:17.4.3`, it just stalls for a couple minutes and then throws the error without attempting to download. (I tried multiple times, all with the same result.) I'm thinking this could be either a timeout because the size of the download is just too big, or it could be an issue with some of the variables not being able to hold a size larger than 2 GB / 4 GB.

When I try to pull `python` (`python:3.11.1`), the download starts but is interrupted part of the way through with a corruption error. (I tried multiple times, all with the same result at close to the same spot: ~17.6 MB downloaded.) This one seems like it might just be either a corrupted file or a bug in the downloader, since it was the only package I had trouble with.

On a good note, I was able to pull `git` and `python:3.11.0` successfully.

nicklauslittle commented:
I'd like to get your thoughts on the Airpower database, @troyaf and @tbateson. I've noticed that the database can get corrupted pretty easily and the only way to fix it is to re-`pull` and manually delete any leftover directories. This corruption can happen for several reasons, including an exception that gets thrown or running multiple commands simultaneously. I know with oldpwr we had the lock file to deal with some of these issues, which I'm assuming we'd like to avoid.

Ideally, I'd like to consolidate the database to only contain necessary information and use the filesystem to look up most of the information. Or maybe even distribute the database to the individual SHA directories stored in something like `$AirpowerHome\content\[SHA]\.pwr` or `$AirpowerHome\content\[SHA]\.pwr_db`. (This would at least allow `Airpower pull` or `Airpower rm` to happen simultaneously as long as the packages were different.)

We might be able to eliminate the database entirely. It seems like most of the information in the database is either redundant or unused:

- `Size` doesn't seem to be used
- `RefCount` only matters if it is `0`, which can be checked using the file system: `(Get-Item "$AirpowerHome/ref/*").Target`
- Link between Tag and Hash can be determined using `(Get-Item "$AirpowerHome/ref/*").Target`

Thoughts?

nicklauslittle commented:
> I'd like to get your thoughts on the Airpower database, @troyaf and @tbateson. I've noticed that the database can get corrupted pretty easily and the only way to fix it is to re-`pull` and manually delete any leftover directories. This corruption can happen for several reasons, including an exception that gets thrown or running multiple commands simultaneously. I know with oldpwr we had the lock file to deal with some of these issues, which I'm assuming we'd like to avoid.
> 
> Ideally, I'd like to consolidate the database to only contain necessary information and use the filesystem to look up most of the information. Or maybe even distribute the database to the individual SHA directories stored in something like `$AirpowerHome\content\[SHA]\.pwr` or `$AirpowerHome\content\[SHA]\.pwr_db`. (This would at least allow `Airpower pull` or `Airpower rm` to happen simultaneously as long as the packages were different.)
> 
> We might be able to eliminate the database entirely. It seems like most of the information in the database is either redundant or unused:
> 
> * `Size` doesn't seem to be used
> * `RefCount` only matters if it is `0`, which can be checked using the file system: `(Get-Item "$AirpowerHome/ref/*").Target`
> * Link between Tag and Hash can be determined using `(Get-Item "$AirpowerHome/ref/*").Target`
> 
> Thoughts?

Here's kind of what I was thinking: https://github.com/nicklauslittle/airpwr/compare/74-pwr..74-pwr-rework-db. It's not complete, but the general idea is there. (Needs quite a bit of cleanup and many tests need to be updated to support it.)

It basically stores a `.pwrdb` file in each `content` directory when it is `untargz`ed (e.g. `{Digest:"sha256:2f59a0baa4215044bea41a569baba96ad575ff9af75409584233a784d520cbc8","Size":40055955}`). (We could probably store additional information in there as well, like maybe unpacked size in addition to the `targz`ed size.) This information is only looked up when it is needed (`Airpower list`, etc.)

The main database keeps roughly the same structure, except that you don't have to save it or worry about it getting out-of-sync with the filesystem, since it is generated from the filesystem metadata. I changed `RefCount` to just `Refs`, which is now a list of directory junctions from the `ref` folder. I also changed the tag map (`{"latest":"sha256:..."}`) so the values are metadata objects with the `Refs` list and the `Directory`.

I wanted to get your thoughts before making any more updates, so please let me know if you have any comments or recommendations.

nicklauslittle-gov commented:
I was trying to use the new Airpower in a CI environment. While trying to pull I was seeing this:

```pwsh
> pwr pull cmake-converter
latest: Pulling cmake-converter
Digest: sha256:cfe7a7b5612f1221c32c50e56256eb20b14456ef6f91ff31398[60](https://github.com/76-eddge/build-images/actions/runs/4105220691/jobs/7081772368#step:3:61)2a89f557a1c
pwr : Exception setting "CursorVisible": "The handle is invalid."
At line:1 char:1
+ pwr pull cmake-converter
+ ~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
```

I'm guessing it's related to the custom progress bar not working properly when running inside CI. Maybe we could have an option to disable the progress bar or not display it when `[console]::CursorVisible = $false`?

nicklauslittle commented:
I'm consistently getting an error when running the following command with the latest `Airpower` on Powershell 7 (5.1 seems to work):

```pwsh
PS > Airpower pull cmake
latest: Pulling cmake
Digest: sha256:e6a11d495297f09149972be6c1adf262a16b7f5347cde434bb5dbe605330444a
6094ffdcbc5d: Extracting █████████████████████████████▌   34.79 MB / 35.18 MB
Invoke-Airpower: Cannot bind argument to parameter 'Buffer' because it is an empty array.
```

troyaf commented:
> I was trying to use the new Airpower in a CI environment. While trying to pull I was seeing this:
> 
> ```powershell
> > pwr pull cmake-converter
> latest: Pulling cmake-converter
> Digest: sha256:cfe7a7b5612f1221c32c50e56256eb20b14456ef6f91ff31398[60](https://github.com/76-eddge/build-images/actions/runs/4105220691/jobs/7081772368#step:3:61)2a89f557a1c
> pwr : Exception setting "CursorVisible": "The handle is invalid."
> At line:1 char:1
> + pwr pull cmake-converter
> + ~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
>     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
> ```
> 
> I'm guessing it's related to the custom progress bar not working properly when running inside CI. Maybe we could have an option to disable the progress bar or not display it when `[console]::CursorVisible = $false`?
> 
> ## Update
> Looks like the error is actually thrown when the line `[Console]::CursorVisible = $false` is hit. Maybe, these lines just need to be disabled when running in CI?

Added a fix but will need to consider whether to allow the progress writing in CI.

troyaf commented:
> > I'd like to get your thoughts on the Airpower database, @troyaf and @tbateson. I've noticed that the database can get corrupted pretty easily and the only way to fix it is to re-`pull` and manually delete any leftover directories. This corruption can happen for several reasons, including an exception that gets thrown or running multiple commands simultaneously. I know with oldpwr we had the lock file to deal with some of these issues, which I'm assuming we'd like to avoid.
> > Ideally, I'd like to consolidate the database to only contain necessary information and use the filesystem to look up most of the information. Or maybe even distribute the database to the individual SHA directories stored in something like `$AirpowerHome\content\[SHA]\.pwr` or `$AirpowerHome\content\[SHA]\.pwr_db`. (This would at least allow `Airpower pull` or `Airpower rm` to happen simultaneously as long as the packages were different.)
> > We might be able to eliminate the database entirely. It seems like most of the information in the database is either redundant or unused:
> > 
> > * `Size` doesn't seem to be used
> > * `RefCount` only matters if it is `0`, which can be checked using the file system: `(Get-Item "$AirpowerHome/ref/*").Target`
> > * Link between Tag and Hash can be determined using `(Get-Item "$AirpowerHome/ref/*").Target`
> > 
> > Thoughts?
> 
> Here's kind of what I was thinking: https://github.com/nicklauslittle/airpwr/compare/74-pwr..74-pwr-rework-db. It's not complete, but the general idea is there. (Needs quite a bit of cleanup and many tests need to be updated to support it.)
> 
> It basically stores a `.pwrdb` file in each `content` directory when it is `untargz`ed (e.g. `{Digest:"sha256:2f59a0baa4215044bea41a569baba96ad575ff9af75409584233a784d520cbc8","Size":40055955}`). (We could probably store additional information in there as well, like maybe unpacked size in addition to the `targz`ed size.) This information is only looked up when it is needed (`Airpower list`, etc.)
> 
> The main database keeps roughly the same structure, except that you don't have to save it or worry about it getting out-of-sync with the filesystem, since it is generated from the filesystem metadata. I changed `RefCount` to just `Refs`, which is now a list of directory junctions from the `ref` folder. I also changed the tag map (`{"latest":"sha256:..."}`) so the values are metadata objects with the `Refs` list and the `Directory`.
> 
> I wanted to get your thoughts before making any more updates, so please let me know if you have any comments or recommendations.

I think you're completely right about how to move forward with the db. The only way to keep it consistent without enforcing crippling synchronicity through a global lock is to use the file system for fine-grained locks. I'll whip up a prototype and take a look at what you wrote.

nicklauslittle commented:
> > I was trying to use the new Airpower in a CI environment. While trying to pull I was seeing this:
> > ```powershell
> > > pwr pull cmake-converter
> > latest: Pulling cmake-converter
> > Digest: sha256:cfe7a7b5612f1221c32c50e56256eb20b14456ef6f91ff31398[60](https://github.com/76-eddge/build-images/actions/runs/4105220691/jobs/7081772368#step:3:61)2a89f557a1c
> > pwr : Exception setting "CursorVisible": "The handle is invalid."
> > At line:1 char:1
> > + pwr pull cmake-converter
> > + ~~~~~~~~~~~~~~~~~~~~~~~~
> >     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
> >     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm guessing it's related to the custom progress bar not working properly when running inside CI. Maybe we could have an option to disable the progress bar or not display it when `[console]::CursorVisible = $false`?
> > ## Update
> > Looks like the error is actually thrown when the line `[Console]::CursorVisible = $false` is hit. Maybe, these lines just need to be disabled when running in CI?
> 
> Added a fix but will need to consider whether to allow the progress writing in CI.

Yeah, it spits out quite a few lines (see https://github.com/76-eddge/build-images/actions/runs/4212235192/jobs/7311113939 for a what it looks like with my hacked Airpower version). I had thought about printing out progress at a slower rate when in CI, but it's not really a necessity. Just maybe a nice-to-have. But, disabling progress completely in CI would also work.

nicklauslittle commented:
> > > I'd like to get your thoughts on the Airpower database, @troyaf and @tbateson. I've noticed that the database can get corrupted pretty easily and the only way to fix it is to re-`pull` and manually delete any leftover directories. This corruption can happen for several reasons, including an exception that gets thrown or running multiple commands simultaneously. I know with oldpwr we had the lock file to deal with some of these issues, which I'm assuming we'd like to avoid.
> > > Ideally, I'd like to consolidate the database to only contain necessary information and use the filesystem to look up most of the information. Or maybe even distribute the database to the individual SHA directories stored in something like `$AirpowerHome\content\[SHA]\.pwr` or `$AirpowerHome\content\[SHA]\.pwr_db`. (This would at least allow `Airpower pull` or `Airpower rm` to happen simultaneously as long as the packages were different.)
> > > We might be able to eliminate the database entirely. It seems like most of the information in the database is either redundant or unused:
> > > 
> > > * `Size` doesn't seem to be used
> > > * `RefCount` only matters if it is `0`, which can be checked using the file system: `(Get-Item "$AirpowerHome/ref/*").Target`
> > > * Link between Tag and Hash can be determined using `(Get-Item "$AirpowerHome/ref/*").Target`
> > > 
> > > Thoughts?
> > 
> > 
> > Here's kind of what I was thinking: https://github.com/nicklauslittle/airpwr/compare/74-pwr..74-pwr-rework-db. It's not complete, but the general idea is there. (Needs quite a bit of cleanup and many tests need to be updated to support it.)
> > It basically stores a `.pwrdb` file in each `content` directory when it is `untargz`ed (e.g. `{Digest:"sha256:2f59a0baa4215044bea41a569baba96ad575ff9af75409584233a784d520cbc8","Size":40055955}`). (We could probably store additional information in there as well, like maybe unpacked size in addition to the `targz`ed size.) This information is only looked up when it is needed (`Airpower list`, etc.)
> > The main database keeps roughly the same structure, except that you don't have to save it or worry about it getting out-of-sync with the filesystem, since it is generated from the filesystem metadata. I changed `RefCount` to just `Refs`, which is now a list of directory junctions from the `ref` folder. I also changed the tag map (`{"latest":"sha256:..."}`) so the values are metadata objects with the `Refs` list and the `Directory`.
> > I wanted to get your thoughts before making any more updates, so please let me know if you have any comments or recommendations.
> 
> I think you're completely right about how to move forward with the db. The only way to keep it consistent without enforcing crippling synchronicity through a global lock is to use the file system for fine-grained locks. I'll whip up a prototype and take a look at what you wrote.

Yeah, I think fine-grained locking is probably the best idea. We could lock on sha256 during download and untar. Then we could lock anytime a DB write is needed, re-read and then re-write the updated DB. Not sure how best to handle multiple concurrent attempts to access the lock. (Maybe print a message and just wait until it's available?)

nicklauslittle commented on src/package.ps1:
```diff
@@ -0,0 +1,485 @@
+	$manifest = $Pkg | ResolveRemoteRef | GetManifest
+	$Pkg.Digest = $manifest | GetDigest
+	$Pkg.Size = $manifest | GetSize
+	WriteHost "$($pkg.Tag | AsTagString): Pulling $($Pkg.Package)"
```
I realize this somewhat matches what `docker pull` prints, but it results in really odd looking strings like `8: Pulling jdk`. Could we change this to something like:
```suggestion
	WriteHost "Pulling $($Pkg.Package):$($pkg.Tag | AsTagString)"
```

troyaf replied:
Yeah

nicklauslittle commented on README.md:
````diff
@@ -1,197 +1,37 @@
-  }
-}
+```PowerShell
+Install-Module Airpower -Scope CurrentUser
````
Not sure if you want to include this, but the following commands can be used to install Airpower without any user prompts (i.e. if you need to run it in a CI job):

```Powershell
Install-PackageProvider -Name NuGet -Force
Set-PSRepository -Name PSGallery -InstallationPolicy Trusted
Install-Module Airpower -Scope CurrentUser
```

troyaf replied:
I think it's worth putting in the README

troyaf commented on src/pwr.ps1:
```diff
@@ -1,1426 +1,260 @@
-	'where' {
-		Assert-NonEmptyPwrPackages
-		foreach ($local:P in $Packages) {
-			$local:Pkg = Assert-PwrPackage $local:P
-			if (-not (Test-PwrPackage $local:Pkg)) {
-				Write-PwrFatal "package $($local:Pkg.Ref) is not installed"
+		$resp = HttpRequest @params | HttpSend -NoRedirect
+		if ($resp.Headers.Location) {
+			$remote = [Version]::new($resp.Headers.Location.OriginalString.Substring('/packages/airpower/'.Length))
+			$local = [Version]::new((Import-PowerShellDataFile -Path "$PSScriptRoot\Airpower.psd1").ModuleVersion)
+			if ($remote -gt $local) {
+				WriteHost "$([char]27)[92mA new version of Airpower is available! [v$remote]$([char]27)[0m"
 			}
-			Write-PwrHost (Resolve-PwrPackage $local:Pkg)
 		}
-	}
-	Default {
-		Write-PwrFatal "no such command '$Command'"
+		PrunePackages -Auto
 	}
 }
```
As written, it runs on module load, so once per shell in which you make an `airpower` call.

If this feels too slow as you use it, we can reinstitute the "daily" cache to prevent this from always running.

nicklauslittle-gov replied:
I wouldn't think this check would be significant, since it only needs to load the local database. But, I'll let you know if I notice anything. I'll also note that I reviewed the implementation, and I'm pretty sure it meets all of my wants for auto-pruning.

nicklauslittle-gov commented:
I noticed an issue with the locking. I was repulling all the packages with the latest Airpower, while accidentally using the jdk with vscode. (That explains the first error, but I would expect it to recover for the remaining commands.)

```powershell
Pulling jdk:latest
Digest: sha256:9e662865b2bac015cdb78809aefb5f8901802c674f7eaa9bdc7e9ff87c9de4d8
28041ebac51b: Extracting                                 128.00 kB / 349.89 MB
pwr : Exception calling "Open" with "3" argument(s): "The process cannot access the file
'\\?\C:\Users\...\AppData\Local\Airpower\content\9e662865b2ba\jdk-17.0.6+10\bin\instrument.dll' because it is being used by another process."
At line:1 char:1
+ pwr exec
+ ~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower

PS C:\Users\...\project> pwr exec
pwr : Exception calling "ReadAllText" with "1" argument(s): "The process cannot access the file
'C:\Users\...\AppData\Local\Airpower\cache\cGtnZGIKamRrCmxhdGVzdA==' because it is being used by another process."
At line:1 char:1
+ pwr exec
+ ~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower

PS C:\Users\...\project> pwr list
pwr : a package pkgdb jdk latest is being used by another airpower process
At line:1 char:1
+ pwr list
+ ~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
```

Deleting the zero length files under the cache directory seemed to fix the issue, but I had to close out of PowerShell before I could delete them. I'm not real familiar with the locking implementation, but I'm guessing we need some `try { } finally { }` blocks?

nicklauslittle-gov commented:
Another issue I've run into with version 0.6.0.17 is that I can't pull several different packages (vs-buildtools, perl, python) even working around the issue above. It downloads and untars the packages just fine, but then I get the following error before the database gets updated. It's always the same directory, which seems odd to me.

```powershell
PS C:\Users\...\project> pwr pull vs-buildtools:17.4.5
Pulling vs-buildtools:17.4.5
Digest: sha256:238cb9dd2f0e045dff5c2635c9e90d91742d44916dad02636f1d7189fdf2c001
92a8fb0baa17: Pull complete
pwr : Exception calling "Delete" with "1" argument(s): "Access to the path 'C:\windows\system32\1033' is denied."
At line:1 char:1
+ pwr pull vs-buildtools:17.4.5
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
```

nicklauslittle-gov commented:
> Another issue I've run into with version 0.6.0.17 is that I can't pull several different packages (vs-buildtools, perl, python) even working around the issue above. It downloads and untars the packages just fine, but then I get the following error before the database gets updated. It's always the same directory, which seems odd to me.
> 
> ```powershell
> PS C:\Users\...\project> pwr pull vs-buildtools:17.4.5
> Pulling vs-buildtools:17.4.5
> Digest: sha256:238cb9dd2f0e045dff5c2635c9e90d91742d44916dad02636f1d7189fdf2c001
> 92a8fb0baa17: Pull complete
> pwr : Exception calling "Delete" with "1" argument(s): "Access to the path 'C:\windows\system32\1033' is denied."
> At line:1 char:1
> + pwr pull vs-buildtools:17.4.5
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
>     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
> ```

I have discovered the reason for this. It appears that any `return` statement is actually equivalent to `Write-Output` (https://devblogs.microsoft.com/powershell/suppressing-return-values-in-powershell-functions/). Since the `GzipRead` function returns the number of bytes read, the return value of `ExtractTarGz` was an array of thousands of ints plus the path at the end. So, `$temp` in `SavePackage` ended up with the int `1033` in it, which happens to exist in `C:\Windows\System32`, which is apparently the default directory for relative paths.

I ended up casting all the uses of `GzipRead` to `void` (e.g. `[void]([Util]::GzipRead($Source, $buf, $Header.Size))`), which solved the problem. I'm putting together a PR.

nicklauslittle-gov commented on src/pwr.ps1:
```diff
@@ -1,1426 +1,265 @@
-			$local:Pkg = Assert-PwrPackage $local:P
-			if (-not (Test-PwrPackage $local:Pkg)) {
-				Write-PwrFatal "package $($local:Pkg.Ref) is not installed"
+		$resp = HttpRequest @params | HttpSend -NoRedirect
```
When loading `Airpower` in an offline environment (starting powershell or `Import-Module Airpower`), this line causes the following error:
```pwsh
Exception calling "GetResult" with "0" argument(s): "An error occurred while sending the request."
At ***\Airpower.psm1:38 char:2
+     $resp = $cli.SendAsync($Req, [Net.Http.HttpCompletionOption]::Res ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : HttpRequestException
```

troyaf replied:
Fixed

nicklauslittle-gov commented:
Experiencing a strange issue while trying to `rm` a package:

```pwsh
PS C:\Users\Me\Code\76cal-test> pwr list

Package         Tag    Digest       Size
-------         ---    ------       ----
jom             latest 19ca3c0e2361 1.27 MB
jdk             11     4cb1aa4abd99 348.79 MB
jdk             8      e39f16178524 193.25 MB
jdk             latest ac9d3f20576f 350 MB
llvm            latest fa4206020c93 815.37 MB
llvm                   abf1ba02c2c4 815.39 MB
node            latest 4debe28e4b76 29.28 MB
nasm            latest db2a58b317e9 509.15 kB
ninja           latest 2f13f35935ab 298.21 kB
powershell      latest 90de93e1481c 103.81 MB
perl            latest a705772c0f18 147.13 MB
python          latest 20fbc35deef2 38.59 MB
python                 520fbdfca6f4 38.38 MB
vs-buildtools   17.4.7 10e709d1a40e 3.99 GB
vs-buildtools   17.6.2 00020d7bf7ac 4.38 GB
vs-buildtools   latest 00020d7bf7ac 4.38 GB
vscode          latest 84a469aefc7f 129.89 MB
vscode                 820ac1ad081c 129.8 MB
vscode                 0de0e64ef725 129.79 MB
7-zip           latest 75fdb92c78cf 2.48 MB
cmake           latest 165323c05980 37.15 MB
cmake                  c62c7aa9b118 35.28 MB
cmake-converter latest b7d550f89cd8 1.39 kB
git             latest b9a240772d1c 143.04 MB
git                    875a59bd9cc7 124.44 MB
gradle          latest 933428453850 118.94 MB
doxygen         latest 94d31b816455 42.46 MB
doxygen                6b70d8abc471 42.08 MB
erlang          latest 9b61e382ef88 128.98 MB


PS C:\Users\Me\Code\76cal-test> pwr rm vs-buildtools:17.4.7
Untagged: vs-buildtools:17.4.7
pwr : Exception calling "Delete" with "2" argument(s): "Could not find a part of the path
'Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.ExpressionCompiler.resources.dll'."
At line:1 char:1
+ pwr rm vs-buildtools:17.4.7
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower

PS C:\Users\Me\Code\76cal-test> pwr list
pwr : a package pkgdb vs-buildtools 17.4.7 is being used by another airpower process
At line:1 char:1
+ pwr list
+ ~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Invoke-Airpower
```

troyaf commented:
> I ended up having to close out of PowerShell, open a new PowerShell, completely remove everything in `C:\Users\Me\AppData\Local\Airpower\ref\vs-buildtools-17.4.7`, and then rerun `pwr rm vs-buildtools:17.4.7`.
> 
> I think there might be 2 separate issues here:
> 
> * The delete call is not working correctly.
> * The locks are not released when an exception occurs. They might need to be wrapped in `try { } finally { }` blocks.

I'll double check the try finallys. I think there is also an outstanding issue with long file paths, so we may have to revert to robocopy again.

I'll see if I can reproduce this in a test first.
